File: app/src/main/java/com/example/data/Appointments.kt
package com.example.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit

/**
 * ميعاد أو مشوار أو التزام غير مالي (`zad_appointments`، ميجريشن 20260914004000).
 * الالتزامات المالية (إيجار/قسط) ليها [ZadObligation] من زمان. العقل بيسجّل المواعيد
 * بـ add_appointment من الشات/تليجرام/المكالمة، وزاد بتفكّر بيها بصوتها قبل ميعادها.
 */
@Serializable
data class ZadAppointment(
    val id: String,
    @SerialName("user_id") val userId: String? = null,
    val title: String,
    val kind: String = "personal",
    @SerialName("starts_at") val startsAt: String,
    @SerialName("place_label") val placeLabel: String? = null,
    @SerialName("remind_minutes_before") val remindMinutesBefore: Int = 30,
    val recurrence: String = "once",
    val status: String = "upcoming",
    val source: String = "app",
    val notes: String? = null,
)

/** قيم `kind` زي ما هي في قيد الجدول — بيانات مطابقة، مش نصوص عرض. */
val APPOINTMENT_KINDS = listOf("work", "errand", "medical", "family", "personal", "other")

enum class AppointmentGroup { TODAY, TOMORROW, THIS_WEEK, LATER, PAST }

fun ZadAppointment.startInstant(): Instant? = try {
    Instant.parse(startsAt)
} catch (_: Exception) {
    try { java.time.OffsetDateTime.parse(startsAt).toInstant() } catch (_: Exception) { null }
}

/**
 * تقسيم المواعيد لمجموعات بتوقيت الجهاز. "اللي فات" = ميعاد وقته عدّى أو اتعلّم خلص —
 * بيتعرض آخر حاجة ومقفول، عشان الشاشة تفضل عن اللي جاي. مواعيد ملغية مابتظهرش.
 * صافية (الوقت والمنطقة داخلين كباراميترز) عشان تتختبر.
 */
fun groupAppointments(
    items: List<ZadAppointment>,
    now: ZonedDateTime,
): List<Pair<AppointmentGroup, List<ZadAppointment>>> {
    val zone: ZoneId = now.zone
    val today = now.toLocalDate()
    val grouped = items
        .filter { it.status != "cancelled" }
        .mapNotNull { a -> a.startInstant()?.let { a to it.atZone(zone) } }
        .sortedBy { it.second }
        .groupBy { (a, at) ->
            val days = ChronoUnit.DAYS.between(today, at.toLocalDate())
            when {
                a.status == "done" || at.isBefore(now) -> AppointmentGroup.PAST
                days == 0L -> AppointmentGroup.TODAY
                days == 1L -> AppointmentGroup.TOMORROW
                days <= 7L -> AppointmentGroup.THIS_WEEK
                else -> AppointmentGroup.LATER
            }
        }
    return AppointmentGroup.entries.mapNotNull { g ->
        grouped[g]?.map { it.first }?.let { list ->
            g to if (g == AppointmentGroup.PAST) list.reversed() else list
        }
    }
}

File: zad_flutter/lib/features/appointments/domain/appointments.dart
/// مواعيدي — Kotlin's `Appointments.kt`: non-money appointments and errands
/// (`zad_appointments`) and reminders tied to a kind of place rather than a
/// time (`zad_place_reminders`). The reminder itself is spoken by the server;
/// nothing here schedules an alarm on the phone.
library;

/// The kinds the table accepts.
const List<String> kAppointmentKinds = <String>[
  'work',
  'errand',
  'medical',
  'family',
  'personal',
  'other',
];

/// The places a place reminder can wait for.
const List<String> kPlaceReminderPlaces = <String>[
  'pharmacy',
  'supermarket',
  'mall',
  'any',
];

/// The repeat choices.
const List<String> kRecurrences = <String>[
  'once',
  'hourly',
  'daily',
  'weekly',
  'monthly',
];

/// One appointment.
typedef Appointment = ({
  String id,
  String title,
  String kind,
  DateTime? startsAt,
  String startsAtRaw,
  String? placeLabel,
  String recurrence,
  String status,
});

/// Reads a row.
Appointment appointmentFromJson(Map<String, dynamic> j) => (
  id: '${j['id']}',
  title: (j['title'] as String?) ?? '',
  kind: (j['kind'] as String?) ?? 'other',
  startsAt: DateTime.tryParse('${j['starts_at']}'),
  startsAtRaw: '${j['starts_at']}',
  placeLabel: j['place_label'] as String?,
  recurrence: (j['recurrence'] as String?) ?? 'once',
  status: (j['status'] as String?) ?? 'upcoming',
);

/// One place reminder.
typedef PlaceReminder = ({String id, String note, String place});

/// Reads a row.
PlaceReminder placeReminderFromJson(Map<String, dynamic> j) => (
  id: '${j['id']}',
  note: (j['note'] as String?) ?? '',
  place: (j['place'] as String?) ?? 'any',
);

/// The list's sections, in order.
enum AppointmentGroup {
  /// Today.
  today('النهارده'),

  /// Tomorrow.
  tomorrow('بكرة'),

  /// Within a week.
  thisWeek('الأسبوع ده'),

  /// Further out.
  later('بعد كده'),

  /// Done or gone.
  past('اللي فات');

  new(this.label);

  /// Heading.
  final String label;
}

/// Kotlin's `groupAppointments`, in [toLocal] civil time: done or already
/// past → past (newest first); otherwise by days from today.
List<(AppointmentGroup, List<Appointment>)> groupAppointments(
  List<Appointment> items,
  DateTime nowLocal,
  DateTime Function(DateTime utc) toLocal,
) {
  final today = DateTime(nowLocal.year, nowLocal.month, nowLocal.day);
  final dated = <(Appointment, DateTime)>[
    for (final a in items)
      if (a.status != 'cancelled' && a.startsAt != null)
        (a, toLocal(a.startsAt!)),
  ]..sort((x, y) => x.$2.compareTo(y.$2));
  final grouped = <AppointmentGroup, List<Appointment>>{};
  for (final (a, at) in dated) {
    final day = DateTime(at.year, at.month, at.day);
    final days = day.difference(today).inDays;
    final g = a.status == 'done' || at.isBefore(nowLocal)
        ? AppointmentGroup.past
        : days == 0
        ? AppointmentGroup.today
        : days == 1
        ? AppointmentGroup.tomorrow
        : days <= 7
        ? AppointmentGroup.thisWeek
        : AppointmentGroup.later;
    (grouped[g] ??= <Appointment>[]).add(a);
  }
  return <(AppointmentGroup, List<Appointment>)>[
    for (final g in AppointmentGroup.values)
      if (grouped[g] case final list?)
        (g, g == AppointmentGroup.past ? list.reversed.toList() : list),
  ];
}

/// Kind names.
String kindLabel(String kind) => switch (kind) {
  'work' => 'شغل',
  'errand' => 'مشوار',
  'medical' => 'دكتور وصحة',
  'family' => 'عيلة',
  'personal' => 'شخصي',
  _ => 'تاني',
};

/// Repeat names.
String recurrenceLabel(String r) => switch (r) {
  'hourly' => 'كل ساعة',
  'daily' => 'كل يوم',
  'weekly' => 'كل أسبوع',
  'monthly' => 'كل شهر',
  _ => 'مرة واحدة',
};

/// Place names.
String placeLabel(String place) => switch (place) {
  'pharmacy' => 'صيدلية',
  'supermarket' => 'سوبرماركت',
  'mall' => 'مول',
  _ => 'أي محل',
};
